From 656dd0e9f719fe7500f26b2bdb40e15a7a1c7f7c Mon Sep 17 00:00:00 2001
From: Sanyog Kale <skale@habana.ai>
Date: Mon, 29 Aug 2022 15:02:33 +0300
Subject: [PATCH] [SW-102303] gaudi3: use timeout expire api in mem init

This patch replaces retry/delay logic with timeout_expired
api in memory init api. Also changed the timeout to 10ms as
per HW team recommendation.

Change-Id: I144e1caa33b9a7e0aca4a46ca84a72f4c9107134
---

diff --git a/arch/gaudi3/arc/common.c b/arch/gaudi3/arc/common.c
index c9bf30c..6de44dd 100644
--- a/arch/gaudi3/arc/common.c
+++ b/arch/gaudi3/arc/common.c
@@ -53,7 +53,7 @@
 
 int arcboot_rom_mem_init(uint64_t addr, uint32_t size)
 {
-	uint32_t retry = HL_INIT_MEM_RETRIES;
+	uint64_t start = arcboot_rom_get_ts_us();
 	uint32_t mem_cfg, cpy_sts;
 
 	regmap_writel((uint32_t)addr, mmD0_PSOC_GLOBAL_CONF_MEM_CPY_DEST_ADDR_L);
@@ -77,11 +77,8 @@
 		if (cpy_sts & D0_PSOC_GLOBAL_CONF_MEM_CPY_STATUS_DONE_M)
 			break;
 
-		retry--;
-		if (!retry)
+		if (arcboot_rom_timeout_expired(start, HL_INIT_MEM_UDELAY))
 			return -1;
-
-		udelay(HL_INIT_MEM_UDELAY);
 	}
 
 	return 0;
diff --git a/include/arcboot.h b/include/arcboot.h
index 8248f77..0710c83 100644
--- a/include/arcboot.h
+++ b/include/arcboot.h
@@ -71,8 +71,7 @@
 
 #define HL_INIT_SP_SRAM_RETRIES			100
 #define HL_INIT_SP_SRAM_UDELAY			10
-#define HL_INIT_MEM_RETRIES			100
-#define HL_INIT_MEM_UDELAY			10
+#define HL_INIT_MEM_UDELAY			10000
 #define SP_SRAM_SIZE_16B_GRANULARITY		0x1000
 #define CRC32_SIZE		4
 #define SYS_REF_CLOCK				50000000ULL
